### PR TITLE
[FLINK-21417][core] Separate type-specific memory segments.

### PR DESCRIPTION
**Pull request by xintongsong**
## What is the purpose of the change

*(For example: This pull request makes task deployment go through the blob server, rather than through RPC. That way we avoid re-transferring them on each deployment (during recovery).)*

## Brief change log

*(for example:)*
  - *The TaskInfo is stored in the blob store on job creation time as a persistent artifact*
  - *Deployments RPC transmits only the blob storage reference*
  - *TaskManagers retrieve the TaskInfo from the blob cache*

## Verifying this change

*(Please pick either of the following options)*

This change is a trivial rework / code cleanup without any test coverage.

*(or)*

This change is already covered by existing tests, such as *(please describe tests)*.

*(or)*

This change added tests and can be verified as follows:

*(example:)*
  - *Added integration tests for end-to-end deployment with large payloads (100MB)*
  - *Extended integration test for recovery after master (JobManager) failure*
  - *Added test that validates that TaskInfo is transferred only once across recoveries*
  - *Manually verified the change by running a 4 node cluser with 2 JobManagers and 4 TaskManagers, a stateful streaming program, and killing one JobManager and two TaskManagers during the execution, verifying that recovery happens correctly.*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / no)
  - The serializers: (yes / no / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / no / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (yes / no / don't know)
  - The S3 file system connector: (yes / no / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / no)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)

### CODE REVIEW

**Comment by xintongsong** [comment]
cc @tillrohrmann

File: flink-core/src/main/java/org/apache/flink/core/memory/MemorySegmentFactory.java
Comment (tillrohrmann) [inline flink-core/src/main/java/org/apache/flink/core/memory/MemorySegmentFactory.java:34]: `A factory for {@link OffHeapMemorySegment}`.

File: flink-core/src/main/java/org/apache/flink/core/memory/OffHeapMemorySegment.java
Comment (tillrohrmann) [inline flink-core/src/main/java/org/apache/flink/core/memory/OffHeapMemorySegment.java:33]: `can be`

File: flink-core/src/test/java/org/apache/flink/core/memory/HybridOffHeapDirectMemorySegmentTest.java
Comment (tillrohrmann) [inline flink-core/src/test/java/org/apache/flink/core/memory/HybridOffHeapDirectMemorySegmentTest.java:54]: The cast should not be necessary if we change the return type of `createSegment`.

File: flink-core/src/test/java/org/apache/flink/core/memory/HybridOffHeapDirectMemorySegmentTest.java
Comment (tillrohrmann) [inline flink-core/src/test/java/org/apache/flink/core/memory/HybridOffHeapDirectMemorySegmentTest.java:52]: Name of the test can be updated.

File: flink-core/src/test/java/org/apache/flink/core/memory/HybridOffHeapDirectMemorySegmentTest.java
Comment (tillrohrmann) [inline flink-core/src/test/java/org/apache/flink/core/memory/HybridOffHeapDirectMemorySegmentTest.java]: I guess this test will be renamed in a later commit, right?

File: flink-core/src/test/java/org/apache/flink/core/memory/HybridOffHeapUnsafeMemorySegmentTest.java
Comment (tillrohrmann) [inline flink-core/src/test/java/org/apache/flink/core/memory/HybridOffHeapUnsafeMemorySegmentTest.java]: Will be renamed later, right?

File: flink-core/src/main/java/org/apache/flink/core/memory/MemorySegmentFactory.java
Comment (tillrohrmann) [inline flink-core/src/main/java/org/apache/flink/core/memory/MemorySegmentFactory.java]: Should we rename the factory methods to reflect which type of off heap memory they allocate?

File: flink-core/src/main/java/org/apache/flink/core/memory/UnsafeMemorySegment.java
Comment (tillrohrmann) [inline flink-core/src/main/java/org/apache/flink/core/memory/UnsafeMemorySegment.java:52]: When is `cleaner` `null`?

File: flink-core/src/test/java/org/apache/flink/core/memory/EndiannessAccessChecks.java
Comment (tillrohrmann) [inline flink-core/src/test/java/org/apache/flink/core/memory/EndiannessAccessChecks.java:45]: could be renamed into `testDirectOffHeapSegment`

File: flink-core/src/test/java/org/apache/flink/core/memory/EndiannessAccessChecks.java
Comment (tillrohrmann) [inline flink-core/src/test/java/org/apache/flink/core/memory/EndiannessAccessChecks.java:50]: Could be renamed into `testUnsafeOffHeapSegment`

**Review by tillrohrmann — APPROVED** [review]
Nice work @xintongsong. I think the changes look good to me. I had a few comments which we should resolve before merging this PR.

File: flink-core/src/main/java/org/apache/flink/core/memory/MemorySegmentFactory.java
Comment (xintongsong) [reply flink-core/src/main/java/org/apache/flink/core/memory/MemorySegmentFactory.java]: Addressed in later commit.

File: flink-core/src/main/java/org/apache/flink/core/memory/UnsafeMemorySegment.java
Comment (xintongsong) [reply flink-core/src/main/java/org/apache/flink/core/memory/UnsafeMemorySegment.java:52]: Nice catch. We should no longer have null cleaner.

**Comment by xintongsong** [comment]
Thanks for the quick response and review, @tillrohrmann.
Comments addressed. I'll merge this once AZP gives green light.

**Pull request closed without merging**
